src/dbus/dbus_thread_object.cpp

```diff
@@ -0,0 +1,449 @@
+}
+
+#if OPENTHREAD_CONFIG_LEGACY_ENABLE
+extern "C" void otNcpHandleDidReceiveNewLegacyUlaPrefix(const uint8_t *aUlaPrefix);
+#endif
+otError DBusThreadObject::SetLegacyULAPrefixHandler(DBusMessageIter &aIter)
+{
+#if OPENTHREAD_CONFIG_LEGACY_ENABLE
+    otIp6Prefix prefix;
+    otError     err = OT_ERROR_NONE;
+
+    VerifyOrExit(DBusMessageExtractFromVariant(&aIter, prefix) == OTBR_ERROR_NONE, err = OT_ERROR_INVALID_ARGS);
+    otNcpHandleDidReceiveNewLegacyUlaPrefix(&prefix.mPrefix.mFields.m8[0]);
+exit:
+    return err;
+#else
+    return OT_ERROR_NOT_IMPLEMENTED;
+#endif // OPENTHREAD_CONFIG_LEGACY_ENABLE
+}
+
+otError DBusThreadObject::SetLinkModeHandler(DBusMessageIter &aIter)
+{
+    auto &           threadHelper = mNcp->GetThreadHelper();
+    otLinkModeConfig cfg;
+    otError          err = OT_ERROR_NONE;
+
+    VerifyOrExit(DBusMessageExtractFromVariant(&aIter, cfg) == OTBR_ERROR_NONE, err = OT_ERROR_INVALID_ARGS);
+    err = otThreadSetLinkMode(threadHelper->GetInstance(), cfg);
+exit:
+    return err;
+}
+
+otError DBusThreadObject::GetLinkModeHandler(DBusMessageIter &aIter)
+{
+    auto &           threadHelper = mNcp->GetThreadHelper();
+    otLinkModeConfig cfg          = otThreadGetLinkMode(threadHelper->GetInstance());
+    otError          err          = OT_ERROR_NONE;
+
+    VerifyOrExit(DBusMessageEncodeToVariant(&aIter, cfg) == OTBR_ERROR_NONE, err = OT_ERROR_INVALID_ARGS);
+exit:
+    return err;
+}
+
+otError DBusThreadObject::GetDeviceRoleHandler(DBusMessageIter &aIter)
+{
+    auto &       threadHelper = mNcp->GetThreadHelper();
+    otDeviceRole role         = otThreadGetDeviceRole(threadHelper->GetInstance());
+    std::string  roleName     = GetDeviceRoleName(role);
+    ;
+    otError err = OT_ERROR_NONE;
+
+    VerifyOrExit(DBusMessageEncodeToVariant(&aIter, roleName) == OTBR_ERROR_NONE, err = OT_ERROR_INVALID_ARGS);
+exit:
+    return err;
+}
+
+otError DBusThreadObject::GetNetworkNameHandler(DBusMessageIter &aIter)
+{
+    auto &      threadHelper = mNcp->GetThreadHelper();
+    std::string networkName  = otThreadGetNetworkName(threadHelper->GetInstance());
+    otError     err          = OT_ERROR_NONE;
+
+    VerifyOrExit(DBusMessageEncodeToVariant(&aIter, networkName) == OTBR_ERROR_NONE, err = OT_ERROR_INVALID_ARGS);
+exit:
+    return err;
+}
+
+otError DBusThreadObject::GetPanIdHandler(DBusMessageIter &aIter)
+{
+    auto &   threadHelper = mNcp->GetThreadHelper();
+    uint16_t panId        = otLinkGetPanId(threadHelper->GetInstance());
+    otError  err          = OT_ERROR_NONE;
+
+    VerifyOrExit(DBusMessageEncodeToVariant(&aIter, panId) == OTBR_ERROR_NONE, err = OT_ERROR_INVALID_ARGS);
+exit:
+    return err;
+}
+
+otError DBusThreadObject::GetExtPanIdHandler(DBusMessageIter &aIter)
+{
+    auto &                 threadHelper = mNcp->GetThreadHelper();
+    const otExtendedPanId *extPanId     = otThreadGetExtendedPanId(threadHelper->GetInstance());
+    uint64_t               extPanIdVal;
```

Comment:
why bother casting to a `uint64_t` here? Can't we encode ByteArray(std::vector<uint8_t>) into a DBus Message?

Reply:
See the discussion here:
https://github.com/openthread/ot-br-posix/pull/360#discussion_r366758462